mcs/class/System.Web/System.Web/HttpApplication.cs

```diff
@@ -1891,24 +1937,99 @@ class AsyncInvoker {
 		public BeginEventHandler begin;
 		public EndEventHandler end;
 		public object data;
+		HttpApplication app;
+		AsyncCallback callback;
 		
-		public AsyncInvoker (BeginEventHandler bh, EndEventHandler eh, object d)
+		public AsyncInvoker (BeginEventHandler bh, EndEventHandler eh, HttpApplication a, object d)
 		{
 			begin = bh;
 			end = eh;
 			data = d;
+			app = a;
+			callback = new AsyncCallback(this.doAsyncCallback);
+			
 		}
 
-		public AsyncInvoker (BeginEventHandler bh, EndEventHandler eh)
-		{
-			begin = bh;
-			end = eh;
-		}
+		public AsyncInvoker (BeginEventHandler bh, EndEventHandler eh, HttpApplication app) : this(bh, eh, app, null) { }
 		
 		public void Invoke (object sender, EventArgs e)
 		{
-			throw new Exception ("This is just a dummy");
+			IAsyncResult res;
+			try {
+			
+				res = begin(this.app, e, this.callback, this.data);
+			}
+			catch(Exception ex) {
+				// completed synchronously because of exception. The Original .NET stuff prohibits void Async operations here,
+				// but I'm not too sure if this applies to mono. At least it shouldn't matter for OWIN.
+				throw;
+			}
+			
+		}
+
+		private void doAsyncCallback(IAsyncResult res)
+		{
+			ThreadPool.QueueUserWorkItem((object ores) => {
+
+				IAsyncResult tres = (IAsyncResult) ores;
+
+				try {
+					end(tres);
+				}
+				catch(Exception ee)
+				{
+					// I tried using ProcessError(), but we only come here frome an Invokation in PipelineDone().
+					// Using ProcessError, I still get a blank screen, this way, we at least log the error to console...
+					Console.Error.WriteLine (ee.ToString());
+				}
+				
+				}, res);
+			
 		}
 	}
 #endregion
+	internal sealed class DynamicModuleManager {
+		private const string moduleNameFormat = "__Module__{0}_{1}";
+
+		private readonly List<DynamicModuleInfo> entries = new List<DynamicModuleInfo>();
+		private bool entriesAreReadOnly = false;
+		private readonly object mutex = new object();
+
+		public void Add (Type moduleType) 
+		{
+			if (moduleType == null) {
+				throw new ArgumentException ("moduleType");
+			}
+			if(!typeof (IHttpModule).IsAssignableFrom(moduleType)) { 
+				throw new ArgumentException ("Given object does not implement IHttpModule.", "moduleType");
+			}
+
+			lock (mutex) {
+				if (entriesAreReadOnly) {
+					throw new InvalidOperationException ("A module was to be added to the dynamic module list, but the list was already initialized. The dynamic module list can only be initialized once.");
+				}
+				entries.Add(new DynamicModuleInfo(moduleType, 
+								string.Format(moduleNameFormat, moduleType.AssemblyQualifiedName, Guid.NewGuid())));
+			}
+		}
+
+		public ICollection<DynamicModuleInfo> LockAndGetModules () {
+			lock (mutex) {
+				entriesAreReadOnly = true;
+				return entries;
+			}
+		}
+	}
+
+	internal struct DynamicModuleInfo {
```

Comment:
'internal' is redundant. Also, since it's a separate top-level type please put it in a separate source file named after the type name